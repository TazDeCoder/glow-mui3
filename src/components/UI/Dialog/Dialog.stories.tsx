import { useState } from 'react';
import type { ComponentMeta, ComponentStory } from '@storybook/react';

import { Dialog } from './Dialog';
import { Button } from '../Button';

export default {
  title: 'Components/UI/Dialog',
  component: Dialog,
} as ComponentMeta<typeof Dialog>;

export const Default: ComponentStory<typeof Dialog> = (args) => {
  const [isMount, setIsMount] = useState(false);

  const openHandler = () => {
    setIsMount((prevState) => {
      if (isMount !== args.open) {
        return !prevState;
      }

      return true;
    });
  };

  const closeHandler = () => {
    setIsMount((prevState) => {
      if (isMount !== args.open) {
        return !prevState;
      }

      return false;
    });
  };

  return (
    <>
      <Button onClick={openHandler}>Open</Button>
      <Dialog
        {...args}
        title="Basic Dialog Title"
        subtitle="Lorem ipsum dolor sit amet, ne reque assum electram vix, vim ferri posse sententiae ut. Est posse facete eu, ex autem pericula scribentur has. At per veritus principes, ut per impetus legendos, cu errem integre admodum qui. Nam discere percipit consequat ad. Ne eum luptatum gubergren intellegam, pro te odio nisl iuvaret."
        open={isMount}
        buttons={[
          { text: 'Cancel', onClick: closeHandler },
          { text: 'Accept', onClick: closeHandler },
        ]}
        onClose={closeHandler}
      />
    </>
  );
};
