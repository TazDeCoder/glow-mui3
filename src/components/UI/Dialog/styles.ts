import type { ThemeUIStyleObject } from '@theme-ui/core';

export const backdropStyles: ThemeUIStyleObject = {
  position: 'fixed',
  top: 0,
  left: 0,
  width: '100vw',
  height: '100vh',
  zIndex: 1201,
  bg: 'scrim',
  opacity: 'backdrop',
};

export const baseStyles: ThemeUIStyleObject = {
  position: 'absolute',
  top: '50%',
  left: '50%',
  zIndex: 1301,
  minWidth: 280,
  maxWidth: 560,
  p: 24,
  bg: 'surface.main',
  borderRadius: 'xl',
  boxShadow: 'lg',
  transform: 'translate(-50%, -50%)',
  '&::before': {
    content: '""',
    position: 'relative',
    top: 0,
    left: 0,
    width: '100%',
    height: '100%',
    bg: 'surface.tint',
    opacity: 'tint',
    borderRadius: 'inherit',
  },
};

export const actionStyles: ThemeUIStyleObject = {
  display: 'flex',
  justifyContent: 'flex-end',
  columnGap: 2,
  pt: 24,
};
