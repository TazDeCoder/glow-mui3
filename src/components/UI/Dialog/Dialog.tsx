/** @jsxRuntime classic */
/** @jsx jsx */

import { forwardRef, HTMLAttributes } from 'react';
import { jsx } from '@theme-ui/core';

import { Box, BoxOwnProps, Portal } from '../../Utils';
import { Typography } from '../Typography';
import { Button } from '../Button';

import { backdropStyles, baseStyles, actionStyles } from './styles';

import type { Assign, ForwardRef, ColorTheme } from '../../types';

type ActionButton = HTMLAttributes<HTMLButtonElement> & {
  text: string;
};

export interface DialogProps
  extends Assign<React.ComponentPropsWithRef<'div'>, BoxOwnProps> {
  /** Component's headline */
  title: string;
  /** Component's supporting text */
  subtitle: string;
  /** Color theme used for the component
   * @default 'primary'
   */
  color?: ColorTheme;
  /** Mount state of the component */
  open: boolean;
  /** Component's action buttons */
  buttons?: ActionButton[];
  /** Handler used when backdrop is clicked  */
  onClose?: () => void;
}

export const Dialog: ForwardRef<HTMLDivElement, DialogProps> = forwardRef(
  (
    {
      title,
      subtitle,
      color = 'primary',
      open,
      buttons,
      sx,
      onClose,
      ...props
    },
    ref,
  ) => {
    if (!open) {
      return null;
    }

    return (
      <Portal>
        <Box sx={backdropStyles} aria-hidden onClick={onClose} />
        <Box
          ref={ref}
          color={`${color}.container`}
          role="dialog"
          aria-labelledby="dialog-title"
          aria-describedby="dialog-description"
          aria-modal
          sx={{
            ...baseStyles,
            ...sx,
          }}
          {...props}
        >
          <Box>
            <Typography
              id="dialog-title"
              as="h2"
              variant="headline"
              size="sm"
              color="surface.onMain"
            >
              {title}
            </Typography>
            <Typography
              id="dialog-description"
              color="surface.onVariant"
              pt={3}
            >
              {subtitle}
            </Typography>
          </Box>
          {buttons && (
            <Box color={`${color}.main`} sx={actionStyles}>
              {buttons.map(({ text, onClick }, idx) => (
                <Button
                  key={idx}
                  color={color}
                  variant="text"
                  onClick={onClick}
                >
                  {text}
                </Button>
              ))}
            </Box>
          )}
        </Box>
      </Portal>
    );
  },
);
