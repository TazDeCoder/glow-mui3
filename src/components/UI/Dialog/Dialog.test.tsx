import { render } from '@testing-library/react';

import { Dialog } from './Dialog';

describe('<Dialog />', () => {
  describe('Snapshots', () => {
    it('renders correctly', () => {
      const { container } = render(<Dialog title="Foo" subtitle="Bar" open />);
      expect(container).toMatchSnapshot();
    });
  });
});
