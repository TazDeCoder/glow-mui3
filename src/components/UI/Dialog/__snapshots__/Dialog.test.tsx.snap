// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Dialog /> Snapshots renders correctly 1`] = `<div />`;
